### PR TITLE
Feature/tidy up mvp

### PR DESCRIPTION
**Pull request by szirbucz**
I created this PR to be able to comment

### CODE REVIEW

File: src/main/java/com/kodekonveyor/tidyup/TidyUser.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import lombok.Data;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Data
+@Entity
+@NoArgsConstructor
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/tidyup/TidyUser.java]: Too many annotations, see my comment at src/main/java/com/kodekonveyor/tidyup/Privilege.java

File: src/main/java/com/kodekonveyor/tidyup/TidyUserController.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	@PostMapping
+	public ResponseEntity<TidyUserResource> post(@Valid @RequestBody final TidyUserDto tidyUserFromRequest) {
+		Optional<TidyUser> already = tidyUserRepository.findByEmail(tidyUserFromRequest.getEmail());
+		if (already.isPresent())
+			throw new TidyUserAlreadyRegisteredException(already.get().getEmail());
+
+		Role role = roleRepository.findByName(tidyUserFromRequest.getRole().toString());
+		TidyUser u = new TidyUser();
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/tidyup/TidyUserController.java]: Don't use 1 char long names.

File: src/main/java/com/kodekonveyor/tidyup/TidyUserDetailsService.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	private TidyUserRepository userRepository;
+	// API
+
+	@Override
+	public UserDetails loadUserByUsername(final String email) throws UsernameNotFoundException {
+		try {
+			final TidyUser user = userRepository.findByEmail(email).map(u -> u)
+					.orElseThrow(() -> new UsernameNotFoundException("No user found with username: " + email));
+
+			return new org.springframework.security.core.userdetails.User(user.getEmail(), user.getPassword(), true,
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/tidyup/TidyUserDetailsService.java]: Fully-qualified name is not needed. Use simply new User.

File: src/main/java/com/kodekonveyor/tidyup/TidyUserDetailsService.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+					true, true, true, getAuthorities(user.getRoles()));
+		} catch (final Exception e) {
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/tidyup/TidyUserDetailsService.java]: What is the reason of the Pokemon exception handling (catch 'em all) here?

File: src/main/java/com/kodekonveyor/tidyup/TidyUserDetailsService.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		final List<String> privileges = new ArrayList<String>();
+		final List<Privilege> collection = new ArrayList<Privilege>();
+		for (final Role role : roles) {
+			collection.addAll(role.getPrivileges());
+		}
+		for (final Privilege item : collection) {
+			privileges.add(item.getName());
+		}
+
+		return privileges;
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/tidyup/TidyUserDetailsService.java]:
```
return roles.stream()
                   .flatMap(Role::getPriviliges)
                   .map(Privilige::getName)
                   .collect(Collectors.toList());
```
Would solve it a bit simpler ;)

File: src/main/java/com/kodekonveyor/tidyup/TidyUserDto.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package com.kodekonveyor.tidyup;
+
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import javax.validation.constraints.Email;
+
+@Getter
+@Setter
+@NoArgsConstructor
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/tidyup/TidyUserDto.java]: No need of NoArgsConstructor. It's implicit in Java until you define a constructor by hand.

File: src/main/java/com/kodekonveyor/tidyup/TidyUserResource.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+import org.springframework.hateoas.ResourceSupport;
+
+import static org.springframework.hateoas.mvc.ControllerLinkBuilder.linkTo;
+import static org.springframework.hateoas.mvc.ControllerLinkBuilder.methodOn;
+
+@Getter
+@NoArgsConstructor
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/tidyup/TidyUserResource.java]: Not needed

File: src/main/java/com/kodekonveyor/tidyup/WorkRequest.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import lombok.Data;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Data
+@Entity
+@NoArgsConstructor
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/tidyup/WorkRequest.java]: Too many annotations, see my comment at src/main/java/com/kodekonveyor/tidyup/Privilege.java

File: src/main/java/com/kodekonveyor/tidyup/WorkRequestDto.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package com.kodekonveyor.tidyup;
+
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@NoArgsConstructor
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/tidyup/WorkRequestDto.java]: Not needed

File: src/main/java/com/kodekonveyor/tidyup/WorkRequestResource.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+public class WorkRequestResource extends ResourceSupport {
+
+	private WorkRequest workRequest;
+
+	public WorkRequestResource(final WorkRequest workRequest) {
+		this.workRequest = workRequest;
+		final long id = workRequest.getId();
+		final long userId = workRequest.getUser().getId();
+		final String city = workRequest.getCity();
+		add(new Link(String.valueOf(id), "request-id"));
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/tidyup/WorkRequestResource.java]: Extract literals to constants.

File: src/main/java/com/kodekonveyor/tidyup/WorkRequestsController.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+	@GetMapping("/workrequests/{city}")
+	public ResponseEntity<Resources<WorkRequestResource>> city(@PathVariable final String city) {
+		final List<WorkRequestResource> collection = getWorkRequestsForCity(city);
+		final Resources<WorkRequestResource> resources = new Resources<>(collection);
+		final String uriString = ServletUriComponentsBuilder.fromCurrentRequest().build().toUriString();
+		resources.add(new Link(uriString, "self"));
+		return ResponseEntity.ok(resources);
+	}
+}
```
Comment (szirbucz) [inline src/main/java/com/kodekonveyor/tidyup/WorkRequestsController.java]: I would split it to multiple controllers.

File: src/main/java/com/kodekonveyor/tidyup/TidyUserResource.java (hunk 7 above)
Comment (grault) [reply src/main/java/com/kodekonveyor/tidyup/TidyUserResource.java]: Removing @Getter, but apparently it seems that serialization needs @NoArgsConstructor (tests fail when I remove that latter annotation).

Comment (magwas) [reply src/main/java/com/kodekonveyor/tidyup/TidyUserResource.java]: There is logic in your explicit constructor which is of course not there in the default one, as there is no implementation given.
Are you sure that it is okay here to have a default constructor without implementation?

I also have philosophical problems with defining one constructor with code, and another with annotation. It is too assymetric. Hence I have a problem with the use of @NoArgsConstructor annotation at all.

File: src/main/java/com/kodekonveyor/tidyup/Privilege.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+import lombok.Data;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Data
+@Entity
+@NoArgsConstructor
```
[the remark this answers is not in the document]
Comment (grault) [reply src/main/java/com/kodekonveyor/tidyup/Privilege.java]: Ack, addressed..

File: src/main/java/com/kodekonveyor/tidyup/Role.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+import lombok.Data;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Data
+@Entity
+@NoArgsConstructor
```
[the remark this answers is not in the document]
Comment (grault) [reply src/main/java/com/kodekonveyor/tidyup/Role.java]: Addressed

File: src/main/java/com/kodekonveyor/tidyup/TidyUser.java (hunk 1 above)
Comment (grault) [reply src/main/java/com/kodekonveyor/tidyup/TidyUser.java]: Agreed, addressed..

File: src/main/java/com/kodekonveyor/tidyup/TidyUserController.java (hunk 2 above)
Comment (grault) [reply src/main/java/com/kodekonveyor/tidyup/TidyUserController.java]: Refactored together with other occurrences.

File: src/main/java/com/kodekonveyor/tidyup/TidyUserDetailsService.java (hunk 3 above)
Comment (grault) [reply src/main/java/com/kodekonveyor/tidyup/TidyUserDetailsService.java]: Yeah, TidyUser used to be called User when practicing.. Forgot about this part of the code after I decided to go with TidyUser instead. Using "new User(" now..

File: src/main/java/com/kodekonveyor/tidyup/TidyUserDetailsService.java (hunk 4 above)
Comment (grault) [reply src/main/java/com/kodekonveyor/tidyup/TidyUserDetailsService.java]: It was not that bad though :) It had a rethrow in it.. I removed this repackaging of exceptions, it indeed looked weird. (it was result of a copy-paste-rewrite originating from questionable stackoverflow sources)

File: src/main/java/com/kodekonveyor/tidyup/TidyUserDetailsService.java (hunk 5 above)
Comment (grault) [reply src/main/java/com/kodekonveyor/tidyup/TidyUserDetailsService.java]: Indeed, I did that.. (flatMap expected to get streams, so I slightly modified the version above)

File: src/main/java/com/kodekonveyor/tidyup/TidyUserDto.java (hunk 6 above)
Comment (grault) [reply src/main/java/com/kodekonveyor/tidyup/TidyUserDto.java]: Addressed

File: src/main/java/com/kodekonveyor/tidyup/WorkRequest.java (hunk 8 above)
Comment (grault) [reply src/main/java/com/kodekonveyor/tidyup/WorkRequest.java]: Ack, done..

File: src/main/java/com/kodekonveyor/tidyup/TidyUserResource.java (hunk 7 above)
Comment (grault) [reply src/main/java/com/kodekonveyor/tidyup/TidyUserResource.java]: I think I get it.. My best guess is in eed2847eb2f541e29360de6f14c5020b19354d2f

File: src/main/java/com/kodekonveyor/tidyup/WorkRequestsController.java (hunk 11 above)
Comment (magwas) [reply src/main/java/com/kodekonveyor/tidyup/WorkRequestsController.java]: Our code complexity metrics are 10 for a method and 20 for a class.
I try to issue a pull request today with automated code quality checks:
- pmd for static analysis (a light one, not containing checks for our implementation pattern),
- unit test coverage, and
- mutation test coverage.

Comment (magwas) [reply src/main/java/com/kodekonveyor/tidyup/WorkRequestsController.java]: The pull request is there.

**Comment by magwas** [comment]
I have addressed the lack of integration and rebuilt the project. Now it fails on compiler warnings. The next will be the pmd violations, and after that the code coverage problems... :)

**Comment by grault** [comment]
Currently CI fails on: "r_kode-konveyor@repository.kodekonveyor.com: Permission denied (publickey)."

**Comment by magwas** [comment]
The old build does not fail for pmd or coverage problems, but we did not allow them. We were using sonar for that, and that would deny the pull request. In the new toolchain the build itself fails. The ssh problem is because the pull request is from outside repo. The new toolchain does not attempt to deliver from a pull request build.

**Comment by grault** [comment]
Ok, so unit tests cover both controllers now..

**Comment by magwas** [comment]
Please accept https://github.com/grault/tidy-up-legacy/pull/1 so the build can fail for the real reasons.

**Comment by magwas** [comment]
Okay, now the shippable build fails for the real reasons.

**Comment by grault** [comment]
Thanks, will look into it later today..

**Comment by magwas** [comment]
Now comes the refactor...

**Comment by magwas** [comment]
For the DTO and Entities (where there is only Lombok generated coe) you can use @ExcludeFromCodeCoverage annotation.

**Comment by grault** [comment]
Sorry for the inactivity, I'll give this request some more attention this week. Thanks for your patience.

**Pull request merged**